### PR TITLE
RATEST-224:Adding Pages for report difinition workflow

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/RATEST-224

### CODE REVIEW

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+   public void clickOnPreviewButton() {
+      clickOn(PREVIEW_BUTTON);
+   }
+
+   public void enterStartDate(String START_DATE) {
+      setTextToFieldNoEnter(START_DATE_FIELD, START_DATE);
+   }
+ 
+   public void enterEndDate(String END_DATE) {
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:28]: even here

File: ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    private static final By EDIT_REPORT_ICON = By.cssSelector("#container > div > table > tbody > tr:nth-child(13) > td:nth-child(6) > a:nth-child(1)");
+
+    public ManageReportsPage(Page page) {
+      super(page);
+    }
+ 
+    public EditReportPage clickOnEditReportPage() {
+      clickOn(EDIT_REPORT_ICON);
+      return new EditReportPage(this);
+    }
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java:15]: The spacing btn line 13 and 14 is the same all through incase that's what you meaning

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+   public EditReportPage(Page parent) {
+      super(parent);
+   }
+
+   public void clickOnPreviewButton() {
+      clickOn(PREVIEW_BUTTON);
+   }
+
+   public void enterStartDate(String START_DATE) {
```
[the remark this answers is not in the document]
Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:24]: > These are static string declarations

Are you sure the variable is static?

> But most openmrs string variables are written in upper case character

Are you able to refer to the documentation that recommends it?

**Comment by kdaud** [comment]
@sherrif10 are you able to have the code indention of [EditReportPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/0514663f9a8473fd280a56b3b26c3dee6ec539de/ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java) and [ManageReportPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/0514663f9a8473fd280a56b3b26c3dee6ec539de/ui-tests/src/test/java/org/openmrs/reference/page/ManageReportsPage.java) similar to the existing code indention within the module?. Among others checkout the indention of [AllergyPage](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/AllergyPage.java)

**Comment by sherrif10** [comment]
cc @kdaud

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	
+	public void clickOnPreviewButton() {
+	    clickOn(PREVIEW_BUTTON);
+	}
+	
+	public void enterStartDate(String startDate) {
+	    setTextToFieldNoEnter(START_DATE_FIELD, startDate);
+	}
+	
+	public void enterEndDate(String endDate) {
```
Comment (sherrif10) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:26]: Here are the references incase [here](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java#L46) and [here](https://github.com/openmrs/openmrs-contrib-qaframework/blob/91cc83e903059428559db923e7dd938f8db31076/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java#L92)

**Comment by kdaud** [comment]
Have left you with two comments before I approve the PR.

However, I must thank you for your resilience and work your doing within the squad. With your tireless effort QA is moving forward and thanks for supporting others folks! I was moved by your compassion with the comment [here ](https://github.com/openmrs/openmrs-contrib-qaframework/pull/45#discussion_r724357808)!

**Comment by sherrif10** [comment]
> ```js
>  homePage->ActiveVisitPage->ClinicianFacingPatientDashboardPage->AllergyPage
> ```

https://qa-refapp.openmrs.org/openmrs/module/reporting/reports/manageReports.form

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	    clickOn(RUN_BUTTON);
+	}
+	
+	public void clickOnCloseTab() {
+	    clickOn(CLOSE_TAB);
+	}
+	
+	@Override
+	public String getPageUrl() {
+	    return "/reporting/reports/reportEditor.form?uuid=b39c4c4c-4881-11e7-a919-92ebcb67fe33";
```
[the remark this answers is not in the document]
Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:40]: @sherrif10 any response on this?

**Comment by kdaud** [comment]
@sherrif10 catch the last comment [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/592#discussion_r729513479) and we finish up the deal!

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:40]: Reports have thier own uuid defined by default, its not generated at run time but rather defined specifically to run that specific report definition, So this means different in built in report have constant defined uuid to use

**Comment by sherrif10** [comment]
@kdaud  responded to your comment above. Thanks

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (dkayiwa) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:40]: Where did you get this uuid from?

**Comment by sherrif10** [comment]
Hello @dkayiwa  is this worth merge

**Comment by kdaud** [comment]
>

@sherrif10 checkout the current post have put in our qa-slack channel and see the opinions there as we look for possible solutions

**Comment by kdaud** [comment]
@sherrif10 the current cucumber report gives outcome for 2.x RefApp workflow tests excluding 3.x tests(this is something to look into once our ci is back to life). cc: @jayasanka-sack 

Based on the cucumber report at https://reports.cucumber.io/reports/7d742f3c-5057-4fd9-b4ac-be9ec7319069 I think issues may be rising from one/two/three of the  3.x RefApp tests whose report is missing on cucumber.io at every end of the build. 

Have also noticed in some builds some of 2.x tests suddenly break which might be due to ther server side but the most persistent report have observed from the several runs have made on the module is the one have shared above.

My suggestion is to disable all the 3.x RefApp tests not to run ci over this weekend(till Sunday) and see the outcome for the build plan.

@sherrif10 @dkayiwa what do you think of my suggestion? or is there any other suggestion that might be better to look into to ensuring CI is green?

**Comment by kdaud** [comment]
> I have temporarily disabled that task as we investigate why.

@dkayiwa does this mean now Bamboo build plans are not running the tests within the module? Or All the tests for 2.x and 3.x are running excluding `publish results to cucumber studio task` !

**Comment by kdaud** [comment]
> @kdaud all tests are running as before. It is only that the very last task of `publish results to cucumber studio` is disabled and hence not run.

Thanks Daniel, have learnt a thing!

Noted with care!

**Comment by dkayiwa** [comment]
FWIW, this is the webhook for that task: https://github.com/openmrs/openmrs-contrib-qaframework/settings/hooks/323453448

**Comment by kdaud** [comment]
> this is the webhook for that task: https://github.com/openmrs/openmrs-contrib-qaframework/settings/hooks/323453448


Worth making an edit in the link otherwise `This is not the web page am looking for`

File: ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	}
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:41]: > ohh the uuid is generate at run time. Have cross checked whether uui remained the same

@sherif10 why not getting a better form url than relying on a generated run time which by default can not be guessed on any random build plan?

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditReportPage.java:41]: @sherrif10 does it mean the report you gave [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/592#discussion_r729654214) was not valid? Or I did not just get the report right?

**Comment by sherrif10** [comment]
Nope it will have different uuid, that's the limitation currently

**Comment by sherrif10** [comment]
The challenge is how will I be able to add a return link that will be equal
to the exact uuid of the report, wanted to use something like getCurrentUrl

**Comment by sherrif10** [comment]
We nolonger this ticket. It was catered for in this pull request   https://github.com/openmrs/openmrs-contrib-qaframework/commit/c92d453434440ab3c7409a233e812478f5260ead#diff-540720529cd53ceeff1514c69d3767c528b73d246b2f4b0f2f9d2b79c82c778c

**Pull request closed without merging**